lib/services/local/userpreferences.go

```diff
@@ -142,36 +146,71 @@ func createBackendItem(username string, preferences *userpreferencesv1.UserPrefe
 	return item, nil
 }
 
-// mergePreferences merges the values from src into dest.
-func mergePreferences(dest, src *userpreferencesv1.UserPreferences) {
-	if src.Theme != userpreferencesv1.Theme_THEME_UNSPECIFIED {
-		dest.Theme = src.Theme
-	}
-
-	if src.Assist != nil {
-		mergeAssistUserPreferences(dest.Assist, src.Assist)
-	}
+// overwriteValues overwrites the values in dst with the values in src.
+func overwriteValues(dst, src protoreflect.ProtoMessage) {
+	d := dst.ProtoReflect()
+	s := src.ProtoReflect()
 
-	if src.Onboard != nil {
-		mergeOnboardUserPreferences(dest.Onboard, src.Onboard)
+	dName := d.Descriptor().FullName().Name()
+	sName := s.Descriptor().FullName().Name()
+	// If the names don't match, then the types don't match, so we can't overwrite.
+	if dName != sName {
+		return
 	}
 
+	overwriteValuesRecursive(d, s)
 }
 
-// mergeAssistUserPreferences merges src preferences into the given dest assist user preferences.
-func mergeAssistUserPreferences(dest, src *userpreferencesv1.AssistUserPreferences) {
-	if src.PreferredLogins != nil {
-		dest.PreferredLogins = src.PreferredLogins
-	}
+// overwriteValuesRecursive recursively overwrites the values in dst with the values in src.
+// It's a helper function for overwriteValues.
+func overwriteValuesRecursive(dst, src protoreflect.Message) {
+	src.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+		switch {
+		case fd.Message() != nil:
+			overwriteValuesRecursive(dst.Mutable(fd).Message(), src.Get(fd).Message())
+		default:
+			dst.Set(fd, src.Get(fd))
+		}
 
-	if src.ViewMode != userpreferencesv1.AssistViewMode_ASSIST_VIEW_MODE_UNSPECIFIED {
-		dest.ViewMode = src.ViewMode
-	}
+		return true
+	})
 }
 
-// mergeOnboardUserPreferences merges src preferences into the given dest onboard user preferences.
-func mergeOnboardUserPreferences(dest, src *userpreferencesv1.OnboardUserPreferences) {
-	if src.PreferredResources != nil {
-		dest.PreferredResources = src.PreferredResources
+// mergeIfUnset merges the values in src with the values in dst if the values in dst are unset.
+func mergeIfUnset(dst, src any) {
+	if reflect.TypeOf(src) != reflect.TypeOf(dst) {
+		return
+	}
+
+	srcV := reflect.ValueOf(src).Elem()
```

Comment:
🙈

Reply:
I know :see_no_evil: But IMO doesn't make sense to have the default values already specified above and keep adding nested ifs to set the same values instead of using reflection here. I was hoping to find something in `proto` package, but `proto.Range()` only iterates over set values, and `proto.Merge()` merges collections, so it won't work here.

Reply:
But we do want to iterate only over fields that are set in `src`, right? So it seems `Range()` should do the job?. OTOH `Range` is pretty close to reflection either way so.. 🤷

Reply:
So.. it can be used? AFAICT the flow is as follow:
- find all fields that are set (non-default) in src
- check if field is set in dst
- if not set it to value from src
`Range` can give you the first step, right?

Reply:
@zmb3 @probakowski I've removed the reflection explicit from our code.